### PR TITLE
jobsite: `manager` and `builder`

### PR DESCRIPTION
**Pull request by supakeen**
WIP, commit messages are not yet complete; TODO below.

Has to be done:

- [ ] Check subdir for export on `builder` side.
- [ ] Comments and documentation (a README probably).
- [ ] Some simple tests.
- [x] CI linters need offerings.
- [x] Add new binaries to RPM spec.
- [x] Disable caching on the builder for osbuild.
- [x] Retry HTTP requests on network errors (but not HTTP errors).

Can be followup:

- [ ] Logs need to be sent over in the `/progress` endpoint on the `builder` and handled in the `manager`.
- [ ] Defining JSON responses for all the routes in the `builder`.
- [ ] Only one export is allowed.
- [ ] How to handle the exports, currently both `-pipeline` and `-export` have to be defined and it feels very awkward?

This PR implements a new concept in `osbuild-composer`: the `jobsite`. The `jobsite` is an isolated set of services to run builds in. They can be separated from the rest of the system through network segmentation or other means. The `jobsite` and what's in it are only available to things inside it. One `jobsite` is used for each build, after which all resources used in it are removed again.

In the case of `osbuild-composer` the worker can set up a network, start a separate virtual machine containing the `builder` and then spawn the `manager` to point at it.

This implements the `jobsite-manager` and `jobsite-builder`. The `builder` is a thin wrapper around `osbuild` itself, it takes in a manifest; build options and allows to exports things.

The `manager` is responsible for sending the manifest and retrieving exports from the `builder`.

Because these commands are potentially running in untrusted environments they are quite picky about what they accept. Both will exit on any kind of error. The `builder` can only be used in the correct order, any out-of-order access will terminate the `builder`. This pickiness extends to file paths not existing or already existing when they're not expected to exist.

## Steps

The `manager` and `builder` know a concept of "states". These are the steps that they both go through in lockstep. The steps are:

1. claim
2. provision
3. populate
4. build
5. progress
6. export

During the **claim**-step the `manager` takes ownership of the `builder`, this is currently a no-op but can contain some jobsite-lifetime secrets. After this the `manager` sends the manifest to the `builder` in the **provision** step. In the **populate** step the `manager` can send any required sources and other inputs to the `builder`, this is currently a no-op but something we want to do soon. The `manager` then instructs the `builder` to start through the **build** step, where it also tells the `builder` what things to export and some other tidbits related to the build.

The `manager` can now start polling the **progress** step. This endpoint will contain logs and progress information as it becomes available during build. Once the **progress** is done we transition into the **export** step where the `manager` retrieves the built artifacts from the `builder`. When this is done both exit and the jobsite can be torn down.

### CODE REVIEW

**Review by croissanne — COMMENTED** [review]
Some small nitpicks, but don't have to address those. The linter errors obviously need to be.

The only thing that's really still missing is to add these binaries to the worker rpm, since that's how the worker images are made.

a small test (with a mock osbuild binary) would also b cool, but  not required to ack this.

**Comment by supakeen** [comment]
@croissanne I've addressed the initial concerns; I'll create a separate commit to add these to the specfile.

**Comment by mvo5** [comment]
Thanks for the PR! This is a very thorough approach that also comes with some complexity. I wonder if we could simplify it a little bit. Part of the complexity seems to be coming from the fact that multiple endpoints have to be called in a specific order and this means a relativ complex API for both client and internal management. If this could be simplfied into a single endpoint that takes all required data it would reduce some of this maybe? 

As an idea maybe we could instead just have a single endpoint, e.g. "/build" that would take a POST multipart/form-data [0]. In there we would have the manifest.json coming as a file, pipelines and environments as form data and (later) sources as a additional (tar) file. Once all that is uploaded the bulder would return a (e.g. randomly generated) URL with the progress of the build for the client. The client then just keeps polling on this endpoint until the result is ready for download. Is there any downside with this simpler approach (we can always add more later if we discover we need it)? Or any other things I'm not considering here?

This is cool stuff! The core is really good, I'm just wondering if we can reduce the complexity of the interface somehow.

[0] It's a bit ugly but it provides what we need and is a standard and can be driven from curl trivially which can be useful for e.g. testing.

File: cmd/osbuild-jobsite-manager/main.go
Comment (ondrejbudai) [inline cmd/osbuild-jobsite-manager/main.go:350]: This needs [`PathEscape`](https://pkg.go.dev/net/url#PathEscape)

**Review by ondrejbudai — COMMENTED** [review]
Michael's suggestion sounds interesting to me, because it can simplify the implementation by a lot. In the end, it's not that different from what's in this PR, but it's just 4 states: build, progress, done and error.

Is there something we would lose by dropping some states? I think that we should consider this API fully private and breakable, so if we need to reintroduce a state, it shouldn't be a big deal.

Some questions/suggestions:
1)  I see that the jobsite builder still has to be subscribed. Are we fine with that for the first iteration?
2) Would be great to document this. :) The PR description is a great start, it can become a README somewhere in the repo.
3) I agree with Sanne, some tests would be nice. A fake manifest touching a file and exporting it should work and be quite simple to write.
4) I'm worried about failing hard on any networking error. I know that this will be running in one cloud region, possibly even in the same AZ. On the other hand, during my work on IB, I saw so many weird network issues trivially fixed with just slapping an http retry, that I wonder if we shouldn't relax the requirements/implementing retries from the beginning. Do we grow the attack surface by implementing retries?

**Comment by supakeen** [comment]
Documentation is indeed sorely lacking; I'll be adding it in this PR and the commit message and PR start are not yet complete. A README is a good idea, perhaps it can be included in the website documentation as well @ochosi?

@mvo5 @ondrejbudai the split came naturally, it's possible to do all of this in a single big request especially since this initial implementation doesn't do much more than that. Perhaps that *is* a better approach.

My idea for splitting them out was to provision some secrets initially and to allow other routes to validate inputs according to the schema without having to receive all the data in a possibly very big and very long-to-transfer request.

File: cmd/osbuild-jobsite-builder/main.go
Comment (supakeen) [reply cmd/osbuild-jobsite-builder/main.go:229]: We can decide on a reasonable `cache-max-size`, @croissanne how much disk are we planning to give these you think? We could even go with no caching initially since we aren't re-using jobsites.

File: cmd/osbuild-jobsite-manager/main.go
Comment (supakeen) [reply cmd/osbuild-jobsite-manager/main.go:350]: Done!

**Comment by ochosi** [comment]
Yup - I can pull in the readme as soon as something has landed. Should be a one-liner🤞 (unless you hide some extra markdown fun for me 😂)

File: cmd/osbuild-jobsite-builder/main.go
Comment (croissanne) [reply cmd/osbuild-jobsite-builder/main.go:229]: The current workers have a 50GB disk, so probably around the same?

Comment (ondrejbudai) [reply cmd/osbuild-jobsite-builder/main.go:244]: I think we should be a bit more defensive here. Checkpoints and caching is not useful at all with ephemeral workers. So I suggest turning it off for now just to keep things minimal. We can change it when we have a proper caching story for jobsites.

Comment (supakeen) [reply cmd/osbuild-jobsite-builder/main.go:244]: Ok.

**Review by croissanne — APPROVED** [review]
:rabbit:

**Comment by achilleas-k** [comment]
> However since we're not currently doing any of that it might be better to remove them again and then replace them as they're being implemented and as needed.

Should we block on this or will it be done in a followup?

**Comment by supakeen** [comment]
> > However since we're not currently doing any of that it might be better to remove them again and then replace them as they're being implemented and as needed.
> 
> Should we block on this or will it be done in a followup?

I don't think we had a concrete outcome if we *actually* want to do that so let's put it in a followup if we want to simplify.

File: cmd/osbuild-jobsite-builder/main.go
Comment (croissanne) [inline cmd/osbuild-jobsite-builder/main.go:244]: Reading this again, it should really run osbuild in `--json` mode, it's needed to parse osbuild's output into an osbuild result.

**Review by croissanne — CHANGES_REQUESTED** [review]
noticed a small thing

File: cmd/osbuild-jobsite-builder/main.go
Comment (supakeen) [reply cmd/osbuild-jobsite-builder/main.go:244]: Done.

**Review by croissanne — APPROVED** [review]

**Pull request merged**
